.github/workflows/equivalence-test.yml

```diff
@@ -1,27 +1,238 @@
+            # Quick sanity check, if the workflow_run that triggered this action
+            # failed then we shouldn't carry on.
+            if [[ "${{ github.event.workflow_run.conclusion }}" != "success" ]]; then
+              echo "::set-output name=run-equivalence-tests::false"
+              exit 0
+            fi
+          
+            # Otherwise, we have to pull our outputs from the workflow_run event 
+            # information.
+            TERRAFORM_VERSION=${{ github.event.workflow_run.head_branch }}
+          
+            if git show-ref -q --verify refs/tags/$TERRAFORM_VERSION; then
+              # Then our trigger was from a new tag being pushed, so we want to
+              # run the equivalence tests and we need to work some things out.
+              RUN_EQUIVALENCE_TESTS=true
+              BUILD_RUN_ID=${{ github.event.workflow_run.id }}
+            else
+              # Then our trigger wasn't from a new tag being pushed, this is
+              # easy as we just skip running the equivalence tests.
+              RUN_EQUIVALENCE_TESTS=false
+            fi
+          fi
+          
+          # One last thing to do is to work out which branch we want to operate
+          # against. This could be `main` for an alpha build, or a release 
+          # branch (eg. v1.1, v1.2, v1.3) for any other kind of build.
+          
+          # Trim the "v" prefix, if any.
+          VERSION="${TERRAFORM_VERSION#v}"
+          
+          # Split off the build metadata part, if any
+          # (we won't actually include it in our final version, and handle it only for
+          # compleness against semver syntax.)
+          IFS='+' read -ra VERSION BUILD_META <<< "$VERSION"
+
+          # Separate out the prerelease part, if any
+          IFS='-' read -r BASE_VERSION PRERELEASE <<< "$VERSION"
+          
+          # Separate out major, minor and patch versions.
+          IFS='.' read -r MAJOR_VERSION MINOR_VERSION PATCH_VERSION <<< "$BASE_VERSION"
+          
+          if [[ "$PRERELEASE" == *"alpha"* ]]; then
+            TARGET_BRANCH=main
+          else
+            TARGET_BRANCH=v${MAJOR_VERSION}.${MINOR_VERSION}
+          fi
+          
+          echo "::set-output name=target-branch::${TARGET_BRANCH}"
+          echo "::set-output name=terraform-version::${TERRAFORM_VERSION}"
+          echo "::set-output name=build-run-id::${BUILD_RUN_ID}"
+          echo "::set-output name=run-equivalence-tests::${RUN_EQUIVALENCE_TESTS}"
+
+  prepare-equivalence-tests:
+    name: "Prepare equivalence testing binary"
+    if: ${{ needs.get-metadata.outputs.run-equivalence-tests == 'true' }}
     runs-on: ubuntu-latest
+    needs:
+      - get-metadata
 
     steps:
-      - run: echo "Hello, world!"
+      - name: "Download terraform-equivalence-testing binary"
+        run: |
+          curl \
+            -H "Accept: application/vnd.github+json" \
+            -H "Authorization: Bearer ${{ secrets.ELEVATED_GITHUB_TOKEN }}" \
```

Comment:
Is there a way we can avoid explicitly issuing long-lived elevated tokens here by instead specifying per-job additional permissions?

For example, I wonder if this job in particular would be able to do its work with "read" access to actions, like this:

```yaml
permissions:
  actions: read
```

The GitHub Actions docs suggest that we can set this as a job-specific setting so that only this particular job has that additional access and the others would only have the `contents: read` access granted by the top-level `permissions`.

If this isn't possible then I'm not opposed to doing this more manually, but if it's possible then I'd rather stick with using the short-lived tokens issued by GitHub Actions itself just to minimize the number of secrets we have in our care.

Reply:
Looking over this again with a broader view just before I submit my comments, I'm realizing that this one in particular is accessing a separate repository and so probably this one unfortunately _does_ need a special token to achieve that.

I was actually intending to leave this comment on the similar line in `prepare-terraform` below, which is accessing an action run in the same repository as this one. I think what I suggested here could potentially work for _that_ case, so at least we could limit the scope of this "elevated token" only to read-only access to the equivalence testing repository. 

(Though maybe we could avoid even _that_ by making `terraform-equivalence-testing` public; seems weird for a job in our public repository to depend on a private repository anyway, and I don't expect there needs to be anything particularly sensitive in that repository, right?)

Reply:
I've removed the ELEVATED token wherever possible and had a chat with Katy about making the other repo public so we can remove the last one. We just kept it private previously so we could avoid any headache that comes with making it opensource (like needing to ensure it's properly licenced and maintained), but now we have a reason to make it open source there's no reason not to.

So I'm leaving this usage of the elevated permissions in place for now, and will come back and remove it once the other repo is public.